Sources/PackageDescription/SupportedPlatforms.swift

```diff
@@ -362,6 +372,28 @@ extension SupportedPlatform {
         @available(_PackageDescription, introduced: 5.3)
         public static let v7: WatchOSVersion = .init(string: "7.0")
     }
+
+    /// The supported WebAssembly features.
+    public struct WasmFeatures: Encodable, OptionSet {
+        public init(rawValue: Int64) {
+            self.rawValue = rawValue
+        }
+
+        /// The underlying features representation.
+        public let rawValue: Int64
+
+        /// Minimum available feature set.
+        static let mvp = WasmFeatures([])
+
+        /// WebAssembly System Interface is available.
+        static let wasi = WasmFeatures(rawValue: 1 << 0)
```

Comment:
I think wasi should be treated as OS in target triple instead of feature flag

Reply:
How do we then distinguish between WASI with atomics and without them? Is WASI expected to be versioned?

Reply:
Or, how do we indicate that these overlap? Say there's a package `A` with `.wasm(.simd)` requirement, but has no assumptions about WASI presence. Then there's another package `B` that assumes presence of WASI, but doesn't require SIMD. A package `C` depends on `A` and `B`. How do we express platform requirements for `C` in its `Package.swift`?

Reply:
WASI is a versioned interface. (see also https://github.com/webassembly/wasi/blob/master/phases/README.md)

> How do we then distinguish between WASI with atomics and without them?

They can be distinguished like below.

```swift
let package = Package(
    name: "SwiftWasmApp",
    platforms: [.wasi(.snapshot, features: [.atomics])]
)

let package = Package(
    name: "SwiftWasmApp",
    platforms: [.wasi(.snapshot)]
)
```

Reply:
Right, and what about packages that don't require WASI, but still need specific Wasm features?

Reply:
We can have two wasm platforms, one for freestanding wasm platform, one for WASI platform.

My opinion is to have two platforms in Package.swift if a user wants to support both platforms.

```swift
let package = Package(
    name: "SwiftWasmApp",
    platforms: [
      .wasi(.snapshot, features: [.atomics]),
      .wasm([.atomics]),
    ]
)
```

I know this syntax is very redundant, but it matches the concept of target triple.

Reply:
Yeah, that makes sense, I'll update the PR accordingly.

Reply:
But I prefer his idea https://forums.swift.org/t/pitch-support-specifying-wasm-features-in-package-manifests/41532/12

> swift build --triple wasm32-unknown-wasi \\
   --experimental-wasm-features wasi,atomics,simd

Reply:
That's an interesting idea, but how would a user know they have to pass this flag to a given package? Especially if the top-level package doesn't need it, but some dependency deep down in the dependency tree does?

Reply:
Users can know that a feature is enabled **at source code level** not **package dependency graph level**.

Reply:
Also abusing the version string is not horrible, but I cast my vote for adding build invocation parameters.

Reply:
Ok, I think that SwiftPM dependency tree checks and source level checks are separate and not mutually exclusive. Other platforms support both. How's about we consider adding source-level checks with the `@available` attribute? https://github.com/apple/swift/pull/34439 is one example of that. We still need to decide how we pass a list of features to it, or if we version Wasm in some way.

For versioning we could consider is year versions. As in, `Wasm 2018` is MVP, `Wasm 2019` is MVP + whatever was available in all browsers at the end of 2019 etc. This would be similar to how Babel versions ECMAScript.

WDYT?

Reply:
Combining the two suggestions, the manifest could indicate:
```swift
let package = Package(
    name: "SwiftWasmApp",
    platforms: [.wasm("atomics,simd")]
```
and SPM translates that into `--experimental-wasm-features`

Reply:
Nice, I like it!